File: java17/src/main/java/com/syd/java17/util/algo/SummaryRanges.java
package com.syd.java17.util.algo;

import java.util.Iterator;
import java.util.Map;
import java.util.TreeMap;

class SummaryRanges {
    TreeMap<Integer, Integer> map = new TreeMap<>();
    public void addNum(int val) {
        Map.Entry<Integer, Integer> floor = map.floorEntry(val), ceil = map.ceilingEntry(val);
        if (floor != null && floor.getValue() >= val) {
            return;
        }
        int key = val, value = val;
        if (floor != null && floor.getValue() == val - 1) {
            key = floor.getKey();
            map.remove(floor.getKey());
        }
        if (ceil != null && ceil.getKey() == val + 1) {
            value = ceil.getValue();
            map.remove(ceil.getKey());
        }
        map.put(key, value);
    }
    public int[][] getIntervals() {
        int n = map.size();
        int[][] res = new int[n][2];
        Iterator<Map.Entry<Integer, Integer>> it = map.entrySet().iterator();
        for (int i = 0; i < n; i++) {
            Map.Entry<Integer, Integer> entry = it.next();
            res[i][0] = entry.getKey();
            res[i][1] = entry.getValue();
        }
        return res;
    }
}

File: kotlin8/src/main/java/com/syd/kotlin8/problem/SummaryRanges.kt
package com.syd.kotlin8.problem

import com.alibaba.fastjson.JSON
import com.syd.kotlin8.invokeResults
import java.util.*


class SummaryRanges() {
    private val map = TreeMap<Int, Int>()
    fun addNum(value: Int) {
        var k = value
        var v = value
        map.floorEntry(value)?.let { (k0, v0) ->
            when (value) {
                in k0..v0 -> return@addNum
                v0 + 1 -> k = k0
            }
        }
        if (map.contains(value + 1)) v = map.remove(value + 1)!!
        map[k] = v
    }

    fun getIntervals(): Array<IntArray> = map.map { intArrayOf(it.key, it.value) }.toTypedArray()
}

fun main() {
    println(JSON.parse(JSON.toJSONBytes(
        invokeResults(SummaryRanges::class,
            "[\"SummaryRanges\",\"addNum\",\"getIntervals\",\"addNum\",\"getIntervals\",\"addNum\",\"getIntervals\",\"addNum\",\"getIntervals\",\"addNum\",\"getIntervals\",\"addNum\",\"getIntervals\",\"addNum\",\"getIntervals\",\"addNum\",\"getIntervals\",\"addNum\",\"getIntervals\",\"addNum\",\"getIntervals\"]",
            "[[],[6],[],[6],[],[0],[],[4],[],[8],[],[7],[],[6],[],[4],[],[7],[],[5],[]]"))))
}

/**
 * Your SummaryRanges object will be instantiated and called as such:
 * var obj = SummaryRanges()
 * obj.addNum(value)
 * var param_2 = obj.getIntervals()
 */